pkg/synchronization/core/ignore.go

```diff
@@ -156,7 +162,65 @@ func (i *ignorer) ignored(path string, directory bool) bool {
 	}
 
 	// Done.
-	return ignored
+	return ignored, false, nil
+}
+
+// dockerIgnorer is a collection of parsed ignore patterns.
+type dockerIgnorer struct {
+	matcher       *fileutils.PatternMatcher
+	parentResults []fileutils.MatchInfo
+}
+
+// newDockerIgnorer creates a new ignorer given a list of user-provided
+// ignore patterns.
+func newDockerIgnorer(patterns []string) (ignorer, error) {
+	ignorePatterns := []string{}
+	for _, p := range patterns {
+		// all docker ignores are absolute, trim off / prefix
+		// if provided
+		if strings.HasPrefix(p, "/") {
+			p = p[1:]
+		}
+
+		ignorePatterns = append(ignorePatterns, p)
+	}
+
+	pm, err := fileutils.NewPatternMatcher(ignorePatterns)
+	if err != nil {
+		return nil, fmt.Errorf("unable to process ignore patterns: %w", err)
+	}
+
+	return &dockerIgnorer{
+		matcher: pm,
+	}, nil
+}
+
+// ignored determines whether or not the specified path should be ignored based
+// on all provided ignore patterns and their order.
+func (i *dockerIgnorer) ignored(path string, directory bool) (bool, bool, error) {
+	if path == "" {
+		return false, false, nil
+	}
+	pathDepth := strings.Count(path, "/")
+
+	var parentResults fileutils.MatchInfo
+	if pathDepth > 0 && len(i.parentResults) >= pathDepth {
+		parentResults = i.parentResults[pathDepth-1]
+	}
+
+	m, matchInfo, err := i.matcher.MatchesUsingParentResults(path, parentResults)
+	if err != nil {
+		return false, false, err
+	}
+	if directory {
+		if len(i.parentResults) > pathDepth {
```

Comment:
Is there a guarantee that this is called as part of an ordered walk? Otherwise, it would be safer to also maintain a stack of parent path components and compare those with the components of `path`.

I see that you reordered the test cases to make them work this way. I don't have great context on the code, but worry that the assumption of ordering might be fragile, even if it happens to hold now.